Add customizable variables for cheatsheet

katomuso opened the pull request:
Add multiple customizable variables to control different aspects of `cider-cheatsheet` and `cider-cheatsheet-select` behavior.

katomuso commented:
@bbatsov Can I be added to the `cider-cheatsheet.el` package authors?

bbatsov commented on cider-cheatsheet.el:
```diff
@@ -34,6 +34,49 @@
+
+(defcustom cider-cheatsheet-auto-select-buffer t
+  "Whether to auto-select the cheatsheet popup buffer."
+  :type 'boolean)
```
You should also add `:package-version` properties to all `defcustoms` (saying `1.15.0` as the version).

katomuso replied:
Done.

bbatsov commented on cider-cheatsheet.el:
```diff
@@ -34,6 +34,49 @@
+                 (const :tag "ClojureDocs" cider-clojuredocs-lookup)
+                 function))
+
+(defcustom cider-cheatsheet-select-path nil
```
Probably it'd better for this to be an option taking some symbol (the name of the type of behavior to use) instead of a boolean - it would make the code more flexible if we decide to add different options down the road.

katomuso replied:
I think that it's a good idea to create a variable like `cider-cheatsheet-select-type` or `cider-cheatsheet-select-display`, which accepts symbols like `multi` (default) and `path`. However, one small consequence of this approach is that we won't be able to use `cider-cheatsheet-select` with a prefix argument anymore as it doesn’t make sense to negate the value of this variable as it did with `cider-cheatsheet-select-path`, which accepted a boolean. Will this be okay?

katomuso replied:
Actually, we can. We just need to check the exact prefix argument value and use the value of `cider-cheatsheet-select-type` accordingly, as it is done in `cider-jack-in-universal`, right?

bbatsov replied:
That's one option, another would be to just drop the prefix argument and rely only on the defcustom.

katomuso replied:
I think it will be better to drop the prefix argument as most users will probably always use one way or the other.

bbatsov commented on cider-cheatsheet.el:
```diff
@@ -34,6 +34,49 @@
+  "Whether to auto-select the cheatsheet popup buffer."
+  :type 'boolean)
+
+(defcustom cider-cheatsheet-doc-function #'cider-doc-lookup
```
Maybe use a more generic name like `cider-cheatsheet-(default)-action-function`? And obviously in the buffer view we can also have more actions bound to different keybindings (the default could be the one triggered by RET)

katomuso replied:
I'm not exactly sure what you mean. Maybe you can point me to some place where such logic is already implemented, so I can understand it better? I think some examples of values for such a variable will also be helpful.

katomuso replied:
If I understand correctly, you're suggesting renaming it to a more generic name because it's not guaranteed that by selecting a var, the user wants to show documentation. Did I understand that correctly?

bbatsov replied:
Yeah, exactly. You can check the namespace navigator or apropos where you can trigger different actions in the buffer displaying the results (e.g. you can show the docs or go to the definition for some var), and there's some default action triggered by `RET`. I'm not sure if that makes much sense here, but it's another case where we are dealing with var names, so we might aim for some consistency.

katomuso replied:
I agree, and it actually looks more generic, which is a good thing.

bbatsov commented on cider-cheatsheet.el:
```diff
@@ -34,6 +34,49 @@
+such as '/' and '&'."
+  :type 'string)
+
+(defcustom cider-cheatsheet-indentation-function
```
I'm not quite certain that many people will want to modify this. Do you have any particular use-case in mind?

katomuso replied:
I thought that, since it is related to UI, some users might want to modify how it looks. However, I agree that most users won't do this, and it is actually not that safe to modify that variable, as it can break `cider-cheatsheet-select`. So, I think it would be a good idea to move it to a constant.

katomuso replied:
Oh, I assumed you meant `cider-cheatsheet-select-path-separator`. If it is `cider-cheatsheet-indentation-function`, then some people prefer 2 spaces, others 4 spaces, tabs, or other indication of indentation. This use-case is what I had in mind when adding this as a customizable variable.

bbatsov replied:
Seems like premature optimization to me. I usually wait for the users to request some config options if I have doubts that there would be demand for them.

katomuso replied:
I think that's a good approach, so for now, they will be removed.

bbatsov commented:
> @bbatsov Can I be added to the cider-cheatsheet.el package authors?

Sure.

katomuso commented:
> > @bbatsov Can I be added to the cider-cheatsheet.el package authors?
> 
> Sure.

Thank you! I'll update the package commentary in a separate PR after this one is finished.

katomuso commented:
I've made all the changes as discussed in the review:

- Renamed `cider-cheatsheet-doc-function` to `cider-cheatsheet-default-action-function` and tweaked docstring
- Renamed `cider-cheatsheet-select-path` to `cider-cheatsheet-select-type`, tweaked docstring and removed prefix argument for `cider-cheatsheet-select`
- Removed `cider-cheatsheet-select-path-separator` and `cider-cheatsheet-indentation-function`

bbatsov commented on cider-cheatsheet.el:
```diff
@@ -57,6 +57,17 @@ from ClojureDocs or any other function accepting a var as an argument."
+If `path`, then represent each completion candidate as a full path to a var."
+  :type '(choice (const multi)
+                 (const path))
+  :package-version '(cider . "0.15.0"))
```
1.15.0 :-)

katomuso replied:
My bad, fixed!

bbatsov commented on cider-cheatsheet.el:
```diff
@@ -34,6 +34,40 @@
+                 function)
+  :package-version '(cider . "0.15.0"))
+
+(defcustom cider-cheatsheet-select-type 'multi
```
Let me think a bit about the names `multi` and `path`, because I'm not sure those are optimal here.

katomuso replied:
Sure. I think so as well, but it was tough to come up with something better.

katomuso replied:
Some alternatives I came up with are 'multi-step and 'full-path, or 'nested and 'flat.

katomuso replied:
Well, come to think of it, maybe just having a prefix argument for `cider-cheatsheet-select` will be enough for now.

bbatsov replied:
Fair enough. :-)

bbatsov commented:
Thanks!
